src/helpers/chain.js

```diff
@@ -0,0 +1,582 @@
+ * @param {object} state oracle's latest state
+ * @param {MonitorMetrics} metrics MonitorMetrics instance
+ * @param {object} amountIn this contains the amountIn for each feed. The feed 
+ *                 name is the field and the value is the amountIn value
+ * @returns {object} last results including the oracle submitted price
+ */
+export const getOracleLatestInfo = async (
+  oracle,
+  oracleDetails,
+  state,
+  metrics,
+  amountsIn
+) => {
+  // Get feeds for oracle
+  let feeds = oracleDetails["feeds"];
+  logger.info("Getting prices for " + oracle + " - " + JSON.stringify(feeds));
+
+  let fromBoard = makeFromBoard();
+  const unserializer = boardSlottingMarshaller(fromBoard.convertSlotToVal);
+  const leader = makeLeader(networkConfig.rpcAddrs[0]);
+
+  const follower = await makeFollower(`:published.wallet.${oracle}`, leader, {
+    unserializer,
+  });
+
+  let offersBalances = await getOffersAndBalances(follower, oracle);
+
+  // Get last offer id from offers from state
+  let lastOfferId = isNaN(state["last_offer_id"]) ? 0 : state["last_offer_id"];
+
+  // Initialise variable to hold results
+  let lastResults = {
+    last_offer_id: lastOfferId,
+    values: state["values"] ? state["values"] : {},
+  };
+
+  // Loop through offers starting from last visited index
+  for (let i = 0; i < offersBalances.offers.length; i++) {
+    // Get current offer
+    let currentOffer = offersBalances.offers[i];
+    let id = Number(currentOffer["status"]["id"]);
+
+    // If we found the last visited offer id in previous check, stop looping
+    if (id <= lastOfferId) {
+      break;
+    }
+
+    // If a price invitation
+    let invMakerName =
+      currentOffer["status"]["invitationSpec"]["invitationMakerName"];
+
+    if (invMakerName === "PushPrice") {
+      let feed =
+        feeds[currentOffer["status"]["invitationSpec"]["previousOffer"]];
+      let lastRound = Number(
+        currentOffer["status"]["invitationSpec"]["invitationArgs"][0]["roundId"]
+      );
+
+      // Get feeds' last observed round from state
+      let lastObservedRound = state["values"].hasOwnProperty(feed)
+        ? state["values"][feed]["round"]
+        : 0;
+
+      // If round is bigger than last observed and the offer didn't fail
+      if (
+        lastRound > lastObservedRound &&
+        !currentOffer["status"].hasOwnProperty("error")
+      ) {
+        // If id is bigger than last offer id in state, set it
+        lastResults["last_offer_id"] = id;
+        lastOfferId = id;
+
+        let price =
+          Number(
+            currentOffer["status"]["invitationSpec"]["invitationArgs"][0][
+              "unitPrice"
+            ]
+          ) / amountsIn[feed];
+
+        // Fill results variable
+        lastResults["values"][feed] = {
+          price: price,
+          id: id,
+          round: lastRound,
+        };
+        state = lastResults;
+
+        // Get latest feed price
+        let feedPrice = await queryPrice(feed);
```

Comment:
if queryPrice fails we return -1 and save it still, is this intentional?

Reply:
Yes, because when a new feed starts, there will not be a price until all oracles submit their first submission and we do not want to exit.